File: backend/app/schemas/webhooks.py
from __future__ import annotations

from datetime import datetime
from typing import Annotated
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field, StringConstraints

from app.models.webhooks import WebhookDeliveryStatus
from app.services.webhooks import WebhookEventType

WebhookName = Annotated[str, StringConstraints(strip_whitespace=True, min_length=1, max_length=120)]
WebhookUrl = Annotated[str, StringConstraints(strip_whitespace=True, min_length=1, max_length=500)]
WebhookSecret = Annotated[
    str,
    StringConstraints(strip_whitespace=True, min_length=1, max_length=255),
]

WEBHOOK_EVENT_TYPES = frozenset(item.value for item in WebhookEventType)


class WebhookEndpointCreate(BaseModel):
    name: WebhookName
    url: WebhookUrl
    secret: WebhookSecret
    event_types: list[str] = Field(min_length=1)

    model_config = ConfigDict(extra="forbid")


class WebhookEndpointRead(BaseModel):
    id: UUID
    name: str
    url: str
    event_types: list[str]
    is_active: bool
    secret_set: bool
    created_by_id: UUID
    created_at: datetime
    updated_at: datetime


class WebhookEndpointListRead(BaseModel):
    items: list[WebhookEndpointRead]
    total: int
    page: int
    page_size: int


class WebhookDeliveryRead(BaseModel):
    id: UUID
    endpoint_id: UUID
    event_id: UUID
    event_type: str
    source_id: str
    payload: dict[str, object]
    status: WebhookDeliveryStatus
    attempt_count: int
    max_attempts: int
    last_error: str | None
    response_status: int | None
    next_retry_at: datetime | None
    last_attempt_at: datetime | None
    delivered_at: datetime | None
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True)


class WebhookDeliveryListRead(BaseModel):
    items: list[WebhookDeliveryRead]
    total: int
    page: int
    page_size: int

File: frontend/src/types/webhooks.ts
export type WebhookEventType =
  | 'project.status_changed'
  | 'payment.created'
  | 'phase.promoted'
  | 'revoke_request.reviewed';

export type WebhookDeliveryStatus =
  | 'pending'
  | 'retry_scheduled'
  | 'delivered'
  | 'dead_letter';

export interface WebhookEndpointRead {
  created_at: string;
  created_by_id: string;
  event_types: WebhookEventType[];
  id: string;
  is_active: boolean;
  name: string;
  secret_set: boolean;
  updated_at: string;
  url: string;
}

export interface WebhookEndpointCreatePayload {
  eventTypes: WebhookEventType[];
  name: string;
  secret: string;
  url: string;
}

export interface WebhookEndpointListQuery {
  page: number;
  pageSize: number;
}

export interface WebhookEndpointListRead {
  items: WebhookEndpointRead[];
  page: number;
  page_size: number;
  total: number;
}

export interface WebhookDeliveryRead {
  attempt_count: number;
  created_at: string;
  delivered_at: string | null;
  endpoint_id: string;
  event_id: string;
  event_type: WebhookEventType;
  id: string;
  last_attempt_at: string | null;
  last_error: string | null;
  max_attempts: number;
  next_retry_at: string | null;
  payload: Record<string, unknown>;
  response_status: number | null;
  source_id: string;
  status: WebhookDeliveryStatus;
  updated_at: string;
}

export interface WebhookDeliveryListQuery {
  page: number;
  pageSize: number;
}

export interface WebhookDeliveryListRead {
  items: WebhookDeliveryRead[];
  page: number;
  page_size: number;
  total: number;
}

export const WEBHOOK_EVENT_OPTIONS: Array<{ label: string; value: WebhookEventType }> = [
  { label: '项目状态变更', value: 'project.status_changed' },
  { label: '付款发生', value: 'payment.created' },
  { label: '环节推进', value: 'phase.promoted' },
  { label: '撤销审核完成', value: 'revoke_request.reviewed' },
];
